Update demographics.py and other files

rickecon opened the pull request:
This PR:
* updates `demographics.py` to use the UN population data portal data for South Africa. This only uses the population, fertility, and mortality data through 2021, which solves Issue #4 . But we could use the forecast data as well, as recommended in Issue #8. However, when I experimented with the forecast data, it looked very different from the past data.
* adds extrapolation functions to `demographics.py` for the fertility rate data. The new functions are ?
* makes minor updates to the `.gitignore`
* updates the `run_og_zaf.py` script to include the new South Africa data in `demographics.py`
* updates the `calibrate.py` file to include the new South Africa data in `demographics.py`
* adds four UN data portal `.csv` files so the function does not have to download them
* updates the `macro_params.py` file to comment out things we don't use and change things we don't have data for. This file still has a bunch of India calibration data. We will do a lot of work here in the next month to update the South Africa calibration.

Another note is that when I run the `run_og_zaf.py` script, the model stops after the first (baseline) steady-state computation because the resource constraint (the equation left out of the solution by Walra's Law) is only solved to 0.0006 precision. That is pretty precise, but it should really be arbitrarily small, like 1e-8. I need to go through the theory and why the resource constraint is not solving precisely. My hypothesis is that there is some inconsistency with how the demographics are handled between the fertility rates, mortality rates, and immigration rates versus the steady-state population distribution. I'll do more work on this. I will leave this PR as a draft (i.e., cannot be merged) until I fix this problem.
```
...
GE loop errors =  [-8.03523914072457e-15, -8.042178034628478e-15, 6.772360450213455e-14, 0.0, 1.3066214776813467e-12, -4.246603069191224e-15, 4.709427292581836e-13, -4.597017211338539e-15, -2.6229018956769323e-15, 4.626715677247262e-13, -3.3757718842508666e-15, -6.262351748276274e-16, 1.175934349895158e-13, -7.314981953499e-14]
Iteration: 01  Distance:  1.5534757137146956e-11
SS debt =  0.6699739181340526 0.006316971073325546
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.1741932187148537
Foreign capital holdings =  0.1631655427856682
resource constraint:  [0.00062941]
Resource Constraint Difference: [0.00062941]
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 131, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 58, in main
    runner(p, time_path=True, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 45, in runner
    ss_outputs = SS.run_SS(p, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/SS.py", line 1159, in run_SS
    output = SS_solver(
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/SS.py", line 902, in SS_solver
    raise RuntimeError(err)
RuntimeError: Steady state aggregate resource constraint not satisfied
```

closes #4

cc: @SeaCelo

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+        pop_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "SexId", "Sex", "AgeMid", "Value"],
```
```suggestion
        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value"],
```
The "AgeMid" is 1 for ages 0 and 1. This may be the reason for the issue in line 123

rickecon replied:
@SeaCelo. Great fix. I made this change.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+        fert_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "AgeMid", "Value"],
```
```suggestion
        usecols=["TimeLabel", "AgeStart", "Value"],
```
The "AgeMid" is 1 for ages 0 and 1. This may be the reason for the issue in line 123

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+            "TimeLabel": "year",
+            "SexId": "sex_num",
+            "Sex": "sex_str",
+            "AgeMid": "age",
```
```suggestion
            "AgeStart": "age",
```

rickecon replied:
@SeaCelo. Done. Thanks.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+    fert_rates_df.rename(
+        columns={
+            "TimeLabel": "year",
+            "AgeMid": "age",
```
```suggestion
            "AgeStart": "age",
```

rickecon replied:
@SeaCelo. Great fix. I made this change.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+    )
+
+    # Clean the data
+    # I don't know why in the pop_df population data by age and sex and year
+    # there are 10 different population numbers for each sex and age and year
+    # and all the other variables are equal. I just average them here.
+    pop_df = (
+        pop_df.groupby(["year", "sex_num", "sex_str", "age"])
+        .mean()
+        .reset_index()
+    )
```
I don't have this problem. I only see a duplicate for ages 0 and 1 and when using the `AgeMid` variable (see proposed changes above to correct this).

rickecon replied:
@SeaCelo. Your `AgeStart` change fixed this problem. Nice catch. I have updated the code such that these lines are deleted because they are no longer necessary.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+        path_folder (None or str): string path to folder where data are stored
+
+    Returns:
+        fert_rates_df (DataFrame): dataset with fertility rates by age
```
```suggestion
        pop_df (DataFrame): dataset with total population by age
```

rickecon replied:
@SeaCelo. Done. Thanks.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -104,83 +360,59 @@ def get_fert(totpers, min_yr, max_yr, graph=False):
-    # Read raw data
-    pop_file = utils.read_file(
-        CUR_PATH, os.path.join("data", "demographic", "india_pop_data.csv")
+    # Get UN fertility rates for India for ages 15-49
```
```suggestion
    # Get UN fertility rates for South Africa for ages 15-49
```

rickecon replied:
@SeaCelo. Good catch. Done.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -104,83 +360,59 @@ def get_fert(totpers, min_yr, max_yr, graph=False):
-            np.float64(max_yr),
-            int(num_sub_bins * max_yr),
+
+    # Extrapolate fertility rates for ages 1-14 and 50-100 using exponential
```
Why are we extrapolating fertility rates for these age groups? The fertility for them is zero, empirically.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -346,13 +515,12 @@ def pop_rebin(curr_pop_dist, totpers_new):
 
 
-def get_imm_resid(totpers, min_yr, max_yr):
+def get_imm_resid(totpers, min_yr, max_yr, graph=False):
```
Why not use the net-migration figures from the UN (indicator 65)?

rickecon replied:
@SeaCelo. I would leave the `graph=False` parameter and default here, just because there is a graphing option. Regarding your question, as we discussed last week, the key is that the population numbers must be consistent. We can either take the population, fertility, and mortality data and back out the immigration rates (as we do here), or we can take the fertility, mortality, and immigration data and use that to get the population data. A third option is to take the population, fertility, mortality, and immigration rate data from the UN database and just hope they are consistent with each other. This is not likely true.

This topic is a little bit nuanced and will take a deeper discussion. But the approach currently taken in `demographics.py` is valid.

SeaCelo replied:
@rickecon Agreed. Though I am convinced the UN Population data is consistent with itself (they also take a residual approach), using the current system works well, so lets keep it as you suggest.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
+    return pop_df
 
 
 def get_fert(totpers, min_yr, max_yr, graph=False):
```
```suggestion
def get_fert(min_yr, max_yr, graph=False):
```
`totpers` is no longer used in the revised function

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,96 +1,352 @@
-    7.Chap_3-Fertility_Indicators-2016.pdf)
 
     Args:
         totpers (int): total number of agent life periods (E+S), >= 3
```
```suggestion
        
```
No longer an argument in the revised function.

rickecon replied:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

rickecon replied:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -104,83 +360,59 @@ def get_fert(totpers, min_yr, max_yr, graph=False):
     return fert_rates
 
 
 def get_mort(totpers, min_yr, max_yr, graph=False):
```
```suggestion
def get_mort(min_yr, max_yr, graph=False):
```

rickecon replied:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

rickecon replied:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -104,83 +360,59 @@ def get_fert(totpers, min_yr, max_yr, graph=False):
+    age. Source: UN Population Data portal.
 
     Args:
         totpers (int): total number of agent life periods (E+S), >= 3
```
```suggestion
        
```

rickecon replied:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

rickecon replied:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -366,44 +534,68 @@ def get_imm_resid(totpers, min_yr, max_yr):
+        .to_numpy()
+        .flatten()
+    )
+    fert_rates = get_fert(totpers, min_yr, max_yr)
+    mort_rates, infmort_rate = get_mort(totpers, min_yr, max_yr)
```
```suggestion
    fert_rates = get_fert(min_yr, max_yr)
    mort_rates, infmort_rate = get_mort(min_yr, max_yr)
```

rickecon replied:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

rickecon replied:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -479,8 +671,7 @@ def get_pop_objs(E, S, T, min_yr, max_yr, curr_year, GraphDiag=False):
-    assert curr_year >= 2011
-    # age_per = np.linspace(min_yr, max_yr, E+S)
+    assert curr_year >= 2021
     fert_rates = get_fert(E + S, min_yr, max_yr, graph=False)
     mort_rates, infmort_rate = get_mort(E + S, min_yr, max_yr, graph=False)
```
```suggestion
    fert_rates = get_fert(min_yr, max_yr, graph=False)
    mort_rates, infmort_rate = get_mort(min_yr, max_yr, graph=False)
```

rickecon replied:
@SeaCelo. I need to work on this a little bit. We discussed hard coding the `totpers` object to be `100` and `min_yr=1` and `max_yr=100`. But I need to look back at all the tests and `calibrate.py` that use this to make sure this is OK. We might need it to be flexible.

rickecon replied:
@SeaCelo. I removed the `min_yr` and `max_yr` arguments in `demographics.py` and `calibrate.py` as they are hard coded to `min_yr=1` and `max_yr=100` because the UN population data have annual ages. I left the `totpers` argument because we need to be able to transform the annual data to model periods that are bigger than one year (`totpers<100`).

SeaCelo reviewed (changes requested):
I made some comments and suggestions to the population, fertility, mortality, and immigration functions.

SeaCelo commented:
@rickecon To help understand the South African population profiles, here are the detailed profiles prepared by UN Population: [https://population.un.org/wpp/Graphs/DemographicProfiles/Line/710](https://population.un.org/wpp/Graphs/DemographicProfiles/Line/710)

SeaCelo commented:
> @rickecon To help understand the South African population profiles, here are the detailed profiles prepared by UN Population: https://population.un.org/wpp/Graphs/DemographicProfiles/Line/710

@rickecon I looked at the population trends for South Africa and the data is correct. South Africa experienced a big change in fertility and mortality due to a few causes. My quick search of the literature give some clues (preferences, risk aversion by women, HIV epidemic and mortality, etc). When we analyze the results, we will need to keep these factors in mind as the population structure has been affected by the large changes in the 2000s.

rickecon commented:
@SeaCelo. When I run the example run script `run_og_zaf.py`, it solves the baseline steady-state equilibrium. However, the transition path equilibrium is converging, but finishes with a resource constraint error. I suspect this is another issue of not getting one of the `demographics.py` time path objects exactly right.
```
Maximum debt ratio:  1.000000025085511
w diff:  5.956415498609147e-07 -6.274392516658622e-08
r diff:  7.951739967870797e-09 -8.710239011999743e-08
r_p diff:  7.951739967870797e-09 -8.710239011999743e-08
p_m diff:  0.0 0.0
BQ diff:  1.6103533946251747e-07 -6.463835558120912e-08
TR diff:  1.5414218641762734e-07 -2.5376482863659344e-07
Iteration: 22
	Distance: 7.272261087114081e-06
Max absolute value resource constraint error: 0.007107331731424233
Checking time path for violations of constraints.
Max Euler error, savings:  7.559286530067766e-12
Max Euler error labor supply:  2.0685675394815917e-12
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 131, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 58, in main
    runner(p, time_path=True, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 68, in runner
    tpi_output = TPI.run_TPI(p, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/TPI.py", line 1293, in run_TPI
    raise RuntimeError(
RuntimeError: Transition path equlibrium not found (RC_error)
```
I will look into this more. For the time being, I am moving this PR back to draft status.

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -104,83 +350,59 @@ def get_fert(totpers, min_yr, max_yr, graph=False):
-    # Read raw data
-    pop_file = utils.read_file(
-        CUR_PATH, os.path.join("data", "demographic", "india_pop_data.csv")
+    # Get UN fertility rates for India for ages 15-49
```
```suggestion
    # Get UN fertility rates for South Africa for ages 15-49
```

rickecon replied:
@SeaCelo. Done. Thanks.

rickecon commented:
@SeaCelo. I put your two comments ([first](https://github.com/EAPD-DRB/OG-ZAF/pull/14#issuecomment-1302562971) and [second](https://github.com/EAPD-DRB/OG-ZAF/pull/14#issuecomment-1303543835)) in Issue #8 so we can work on those in the future. That is something that we want to treat outside of this PR.

rickecon commented:
@SeaCelo. I made the changes that remove the `min_yr` and `max_yr` arguments from the functions in the `demographics.py` module. These are now hard coded to be `min_yr=1` and `max_yr=100` to correspond with the annual age data in the UN population data. I left the `totpers` arguments. This allows the model to adjust from annual model years `totpers=100` to model years that are fewer `totpers<100`. This is valuable in testing situations and other situations where you want to run a smaller and quicker version of the model.

The steady-state still solves, but I am getting the same resource constraint error in the baseline time path after it converges.
```
GE loop errors =  [-1.1796119636642288e-16, -1.249000902703301e-16, 1.5543122344752192e-15, 0.0, 1.341149413747189e-13, 3.8510861166685117e-16, -1.0408340855860843e-15, -1.2385925618474403e-15, -6.782768791069316e-16, -7.73686670285656e-16, -1.0130785099704553e-15, -4.3194614551822497e-16, 1.2073675392798577e-14, -8.590350653037149e-15]
Iteration: 01  Distance:  1.2025259571510712e-13
SS debt =  0.6693507845437849 0.006379343173246392
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.17403120398138408
Foreign capital holdings =  0.17525777572767656
resource constraint:  [-2.39946951e-14]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.6564527527407336e-13
Maximum error in savings FOC =  7.949196856316121e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_BASELINE/SS/SS_vars.pkl
```
The transition path error output after running the baseline is the following:
```
Maximum debt ratio:  1.000000023879215
w diff:  5.778020204516565e-07 -5.8068007069422833e-08
r diff:  7.361659268290133e-09 -8.447208207224e-08
r_p diff:  7.361659268290133e-09 -8.447208207224e-08
p_m diff:  0.0 0.0
BQ diff:  1.5559656061162053e-07 -4.9671494802699945e-08
TR diff:  1.5398244257447402e-07 -2.456320713772886e-07
Iteration: 22
	Distance: 7.07245418554592e-06
Max absolute value resource constraint error: 0.007064895621105965
Checking time path for violations of constraints.
Max Euler error, savings:  7.44737604918555e-12
Max Euler error labor supply:  1.9104717807749694e-12
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 131, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 58, in main
    runner(p, time_path=True, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 68, in runner
    tpi_output = TPI.run_TPI(p, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/TPI.py", line 1293, in run_TPI
    raise RuntimeError(
RuntimeError: Transition path equlibrium not found (RC_error)
```
I'll take this PR out of draft status after I fix this error.

rickecon commented:
@SeaCelo. I fixed one error in `demographics.py` but the resource constraint in the transition path equilibrium is still not solving. The current error I fixed, which I thought would solve the whole problem, was a typo in the `get_imm_resid()` function in `demographics.py`. I had to fix the denominator of the lines (lines 727-732) that compute the immigration rate as a residual. It used to be `pop_2020_EpS[1:]`, but it needed to be `pop_2020_EpS[:-1]`. Dumb typo.
```python
imm_rates_s_2020 = (
    pop_2021_EpS[1:] - (1 - mort_rates[:-1]) * pop_2020_EpS[:-1]
) / pop_2020_EpS[:-1]
imm_rates_s_2019 = (
    pop_2020_EpS[1:] - (1 - mort_rates[:-1]) * pop_2019_EpS[:-1]
) / pop_2019_EpS[:-1]
```
I did have to change the starting interest rate to `initial_guess_r_SS = 0.055` in `ogzaf_default_parameters.json`. As with previous versions of this PR, the steady-state solves with all equations, including the resource constraint, solving to arbitrary accuracy.
```
GE loop errors =  [-4.506811590587745e-14, -4.508893258758917e-14, 3.2440716779547074e-13, 0.0, -6.094347249074872e-12, 1.477290512141849e-14, 4.3770542745846797e-14, -9.944371714976086e-13, 5.0964441000722616e-14, 6.760217385881617e-14, 9.533346334578141e-14, 2.1935578353726726e-14, -5.484918075282508e-13, -1.179903397208193e-12]
Iteration: 01  Distance:  2.529198625454508e-11
SS debt =  0.6711779592951316 0.006393461432212574
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.17450626941673422
Foreign capital holdings =  0.20186988283166193
resource constraint:  [-6.64399091e-14]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.5276668818842154e-13
Maximum error in savings FOC =  7.238654120556021e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_BASELINE/SS/SS_vars.pkl
```
But the transition path equilibrium still does not solve the resource constraint to arbitrary precision. The solution method converges, but the resource constraint error in the converged problem is 0.005, which is too big.
```
Maximum debt ratio:  1.000000021248
w diff:  5.317299547336063e-07 -6.817061692920845e-08
r diff:  8.724744218757685e-09 -7.817221110012484e-08
r_p diff:  8.724744218757685e-09 -7.817221110012484e-08
p_m diff:  0.0 0.0
BQ diff:  1.409597274459884e-07 -5.723262022291897e-08
TR diff:  1.733481109628121e-07 -2.2378731083516623e-07
Iteration: 22
	Distance: 6.44583182750035e-06
Max absolute value resource constraint error: 0.0054670522852872366
Checking time path for violations of constraints.
Max Euler error, savings:  7.541522961673763e-12
Max Euler error labor supply:  1.9180212973424204e-12
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 131, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 58, in main
    runner(p, time_path=True, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 68, in runner
    tpi_output = TPI.run_TPI(p, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/TPI.py", line 1293, in run_TPI
    raise RuntimeError(
RuntimeError: Transition path equlibrium not found (RC_error)
```
I will keep working on this.

rickecon commented:
@SeaCelo. I fixed an error in the `demographics.py` module in the `get_pop_objs()` function in which the steady-state population growth rate `g_n_ss` was based on the original population law of motion with the original immigration rates. This change made `g_n_ss` based correctly on the adjusted immigration rates. The steady-state solves.
```
GE loop errors =  [1.3584827707191494e-12, 1.3584897096130533e-12, -1.3342438265340206e-11, 0.0, -6.299083477045997e-11, -4.4683146072088675e-12, 3.1430170965851545e-12, 3.014186122918261e-12, 2.113132585579436e-12, -4.1530529015787465e-12, -1.0508271336417963e-11, 1.2577005409353248e-12, -5.66917346400686e-12, 6.457390178127298e-12]
Iteration: 01  Distance:  2.8559618457782164e-10
SS debt =  0.6711779592859342 0.006393461432124962
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.17450626941434288
Foreign capital holdings =  0.20186988275228981
resource constraint:  [-4.47763354e-13]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.5631940186722204e-13
Maximum error in savings FOC =  6.572520305780927e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_BASELINE/SS/SS_vars.pkl
```
But the transition path still breaks after converging because the resource constraint is not solved to arbitrary precision.
```
Maximum debt ratio:  1.0000000212422069
w diff:  5.317299542895171e-07 -6.817061737329766e-08
r diff:  8.724744232635473e-09 -7.817221105155259e-08
r_p diff:  8.724744239574367e-09 -7.817221105155259e-08
p_m diff:  0.0 0.0
BQ diff:  1.4095972736272167e-07 -5.723262049006639e-08
TR diff:  1.7334811101832326e-07 -2.2378731083516623e-07
Iteration: 22
	Distance: 6.445831828830127e-06
Max absolute value resource constraint error: 0.005467052285404431
Checking time path for violations of constraints.
Max Euler error, savings:  7.496225862269057e-12
Max Euler error labor supply:  1.9104717807749694e-12
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 131, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 58, in main
    runner(p, time_path=True, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 68, in runner
    tpi_output = TPI.run_TPI(p, client=client)
  File "/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/TPI.py", line 1293, in run_TPI
    raise RuntimeError(
RuntimeError: Transition path equlibrium not found (RC_error)
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        pop_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value"],
+        float_precision="round_trip",
+    )
```
```suggestion
        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value", "VariantId"],
        float_precision="round_trip",
    )
    pop_df=pop_df.loc[pop_df['VariantId'] == variant]
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        fert_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "AgeStart", "Value"],
+        float_precision="round_trip",
+    )
```
```suggestion
        usecols=["TimeLabel", "AgeStart", "Value", "VariantId"],
        float_precision="round_trip",
    )
    fert_rates_df=fert_rates_df.loc[fert_rates_df['VariantId'] == variant]
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+    country_id: str = "710",
+    start_year: int = 2021,
+    end_year: int = None,
+    download: bool = True,
+) -> pd.DataFrame:
```
```suggestion
    download: bool = True,
    variant: int = 4,
) -> pd.DataFrame:
```
Note: this is only needed for the population data. The deaths data is only reported for the Median variant

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        country_id (str): 3-digit country id (numerical)
+        start_year (int): beginning year of the data
+        end_year (int): end year of the data
+        download (bool): whether to download the data from the UN Data Portal.
+            If False, a path must be specified in the path_folder argument.
```
```suggestion
        download (bool): whether to download the data from the UN Data Portal.
            If False, a path must be specified in the path_folder argument.
        variant (int): the model variant used to generate forecasts. Default is "median" (4).
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        pop_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value"],
+        float_precision="round_trip",
+    )
```
```suggestion
        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value", "VariantId"],
        float_precision="round_trip",
    )
    pop_df=pop_df.loc[pop_df['VariantId'] == variant]
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+    country_id: str = "710",
+    start_year: int = 2021,
+    end_year: int = None,
+    download: bool = True,
+) -> pd.DataFrame:
```
```suggestion
    download: bool = True,
    variant: int = 4,
) -> pd.DataFrame:
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        country_id (str): 3-digit country id (numerical)
+        start_year (int): beginning year of the data
+        end_year (int): end year of the data
+        download (bool): whether to download the data from the UN Data Portal.
+            If False, a path must be specified in the path_folder argument.
```
```suggestion
        download (bool): whether to download the data from the UN Data Portal.
            If False, a path must be specified in the path_folder argument.
        variant (int): the model variant used to generate forecasts. Default is "median" (4).
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        pop_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value"],
+        float_precision="round_trip",
+    )
```
```suggestion
        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value", "VariantId"],
        float_precision="round_trip",
    )
    pop_df=pop_df.loc[pop_df['VariantId'] == variant]
```

SeaCelo reviewed (commented):
I added code that limits the data returned from UN Population API for a specified model variant. The default is median model (=4).

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+            + str(end_year)
+            + "?format=csv"
+        )
+        infmort_target = (
+            "https://population.un.org/dataportalapi/api/v1/data/indicators/"
+            + infmort_code
+            + "/locations/"
+            + country_id
+            + "/start/"
+            + str(start_year)
+            + "/end/"
+            + str(end_year)
+            + "?format=csv"
+        )
```
#19 has code to get the neonatal mortality value directly

SeaCelo replied:
```suggestion
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        usecols=["TimeLabel", "SexId", "Sex", "AgeStart", "Value"],
+        float_precision="round_trip",
+    )
+    infmort_rate_df = pd.read_csv(
+        infmort_target,
+        sep="|",
+        header=1,
+        usecols=["TimeLabel", "SexId", "Sex", "Value"],
+        float_precision="round_trip",
+    )
```
#19 updates this

SeaCelo replied:
```suggestion

```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+    mort_rates_df = deaths_df.copy()
+
+    # Clean the data
+    infmort_rate_df["infmort_rate"] = (
+        infmort_rate_df["inf_deaths_p_1000"] / 1000
+    )
+
+    infmort_rate_df = infmort_rate_df[
+        (
+            (infmort_rate_df["year"] >= start_year)
+            & (infmort_rate_df["year"] <= end_year)
+        )
+    ]
```
#19 updates the source of infmort_rate_df

SeaCelo replied:
```suggestion

```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -195,111 +491,120 @@ def get_mort(totpers, min_yr, max_yr, graph=False):
+        raise ValueError(err_msg)
+
+    # Get UN infant mortality and mortality rate data by age
+    infmort_rate_df, mort_rates_df = get_un_mort_data(
+        start_year=2021, download=False
```
#19 updates this

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
 from ogcore import parameter_plots as pp
-
+import matplotlib.pyplot as plt
 
```
```suggestion
from pandas_datareader import wb
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+) -> pd.DataFrame:
+    """
+    Get UN mortality rate data for a country for some range of years (at least
+    one year) and by age, and get infant mortality rate data. The
```
```suggestion
    one year) and by age. The
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        path_folder (None or str): string path to folder where data are stored
+
+    Returns:
+        infmort_rate_df (DataFrame): dataset with infant mortality rates by yr
```
```suggestion

```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -195,111 +491,120 @@ def get_mort(totpers, min_yr, max_yr, graph=False):
-                73.6,
-                104.8,
-                167.6,
+    infmort_rate = infmort_rate_df["infmort_rate"][
+        infmort_rate_df["sex_num"] == 3
+    ].to_numpy()[0]
```
```suggestion
infmort_rate = wb.download(indicator="SH.DYN.NMRT", country="ZAF", start=2020, end=2020).squeeze()/1000

```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        )
+    else:
+        pop_target = os.path.join(DATA_DIR, "un_zaf_pop.csv")
+        infmort_target = os.path.join(DATA_DIR, "un_zaf_infmort.csv")
```
```suggestion
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        },
+        inplace=True,
+    )
+    infmort_rate_df.rename(
+        columns={
+            "TimeLabel": "year",
+            "SexId": "sex_num",
+            "Sex": "sex_str",
+            "Value": "inf_deaths_p_1000",
+        },
+        inplace=True,
+    )
```
```suggestion
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -1,192 +1,488 @@
+        )
+    ]
+
+    return infmort_rate_df, mort_rates_df
```
```suggestion
    return mort_rates_df
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -195,111 +491,120 @@ def get_mort(totpers, min_yr, max_yr, graph=False):
+        raise ValueError(err_msg)
+
+    # Get UN infant mortality and mortality rate data by age
+    infmort_rate_df, mort_rates_df = get_un_mort_data(
```
```suggestion
    mort_rates_df = get_un_mort_data(
```

SeaCelo commented on ogzaf/demographics.py:
```diff
@@ -321,89 +626,120 @@ def pop_rebin(curr_pop_dist, totpers_new):
+    pop_2021_EpS = pop_rebin(pop_2021, totpers)
+
+    fert_rates = get_fert(totpers)
+    mort_rates, infmort_rate = get_mort(totpers)
+
```
```suggestion
    mort_rates = get_mort(totpers)
    infmort_rate = wb.download(indicator="SH.DYN.NMRT", country="ZAF", start=2020, end=2020).squeeze()/1000

```

rickecon commented:
@SeaCelo. I fixed the issue with the transition path equilibrium not satisfying the resource constraint in the first period. For the transition path, we have to compute the population distribution in the period before the initial period. Things like the total capital stock and total bequests in the first period depend on the population distribution in the period before that. So `demographics.py` produces an `omega` object that is the population distribution from period `t=1` to period `t=T+S`. The module also produces `omega_S_preTP`, which is the population distribution in period `t=0`. My error was that I was taking that data from the actual data. However, the data are not consistent with the population demographics law of motion we specified in the theory, at least not with constant fertility rates and immigration rates. For this reason, I had to make `omega_S_preTP` consistent with the `omega` matrix (in terms of fertility rates, mortality rates, and immigration rates).

The baseline steady-state solution output is the following:
```
GE loop errors =  [-8.227446501862801e-14, -4.869021852371702e-14, -1.3922196728799463e-13, 0.0, 2.1715962361668062e-13, -1.293652684974944e-13, -5.678790770957676e-14, -5.6534638082084143e-14, -6.17214612752548e-15, -3.7567171595753734e-14, -5.0067588963642606e-14, -1.5612511283791264e-14, 1.9546864127306662e-14, 3.039235529911366e-15]
Iteration: 01  Distance:  4.784789093966305e-12
SS debt =  0.6711715191110103 0.0063955512984892274
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.17450459496886267
Foreign capital holdings =  0.20239138201602544
resource constraint:  [-4.2802567e-14]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.6076029396572267e-13
Maximum error in savings FOC =  7.638334409421077e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_BASELINE/SS/SS_vars.pkl
```

The baseline transition path solution output is the following;
```
Maximum debt ratio:  1.0000000212110733
w diff:  5.364696991083662e-07 -7.536044477340909e-08
r diff:  9.654272145609433e-09 -7.871194662723013e-08
r_p diff:  9.654272145609433e-09 -7.871194662723013e-08
p_m diff:  0.0 0.0
BQ diff:  1.4075238878932805e-07 -5.199504206701566e-08
TR diff:  1.678670771831925e-07 -2.235329153860155e-07
Iteration: 22
	Distance: 6.4398646151337245e-06
Max absolute value resource constraint error: 1.0536127898710945e-05
Checking time path for violations of constraints.
Max Euler error, savings:  7.473133223356854e-12
Max Euler error labor supply:  1.9113599591946695e-12
Time path iteration complete.
It took 1815.7290947437286 seconds to get that part done.
run time =  1815.729147195816
```

The reform steady-state solution output is the following:
```
GE loop errors =  [7.81388842518993e-14, 8.174017018802715e-14, -4.771738559838923e-13, 0.0, 3.5582647939236267e-13, -2.2105581254372453e-14, -1.0152295670806666e-13, -3.642919299551295e-16, -2.207435623180487e-14, 4.0554365421385796e-14, -2.8102520310824275e-16, 2.6020852139652106e-18, 3.202299536653186e-14]
Iteration: 01  Distance:  3.1749021791539607e-12
SS debt =  1.2004356041121698 0.015262156036278306
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.48017424164486794
Foreign capital holdings =  0.08048294759663217
resource constraint:  [1.7069679e-14]
Steady state government spending is negative to satisfy budget
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  1.3011813848606835e-13
Maximum error in savings FOC =  7.149836278586008e-14
/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/SS.py:1319: UserWarning: Warning: The combination of the tax policy you specified and your target debt-to-GDP ratio results in an infeasible amount of government spending in order to close the budget (i.e., G < 0)
  warnings.warn(
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_REFORM/SS/SS_vars.pkl
```

And the reform transition path solution output is the following:
```
Maximum debt ratio:  2.000104723711181
w diff:  2.5915477188664937e-06 -1.1417283154457891e-06
r diff:  1.3815679607981357e-07 -4.063585497746103e-07
r_p diff:  8.413844220650502e-08 -2.596077518443596e-07
p_m diff:  0.0 0.0
BQ diff:  1.1468624764796043e-07 -5.7201863237418826e-08
TR diff:  3.424993502282314e-07 -2.4845994033712016e-07
Iteration: 21
	Distance: 9.290020424167063e-06
Max absolute value resource constraint error: 6.777817169005551e-07
Checking time path for violations of constraints.
Max Euler error, savings:  9.631240249774464e-12
Max Euler error labor supply:  1.979305608301729e-12
Time path iteration complete.
It took 1644.864609003067 seconds to get that part done.
run time =  1644.8646562099457
Percentage changes in aggregates: Year                    Variable  2021  2022  ...  2030  2021-2030     SS
0                    GDP ($Y_t$)  3.71  2.09  ... -2.05      -0.58 -10.57
1            Consumption ($C_t$) 16.87  7.72  ... 12.41      12.78   6.61
2          Capital Stock ($K_t$)  1.97 -0.80  ... -5.58      -4.45 -26.04
3                  Labor ($L_t$)  0.57 -0.38  ... -4.01      -2.39  -4.80
4     Real interest rate ($r_t$)  9.35 12.07  ... 14.06      14.70  52.48
5                      Wage rate  1.05  0.41  ... -0.00      -0.17  -7.95

[6 rows x 13 columns]
```

SeaCelo commented:
@rickecon I can't checkout this  branch.

`git checkout -b rickecon-demographics`
`git pull https://github.com/rickecon/OG-ZAF.git demog`

Gives me an error: 
```
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (14/14), done.
remote: Total 29 (delta 14), reused 14 (delta 14), pack-reused 15
Unpacking objects: 100% (29/29), 14.20 KiB | 632.00 KiB/s, done.
From https://github.com/rickecon/OG-ZAF
 * branch            demog      -> FETCH_HEAD
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.

```

SeaCelo commented:
I was able to successful run the model. 
FYI, these values for `r_guess` have a SS solution: 0.02, 0.035, 0.04, 0.06, 0.065, 0.07, 0.075, 0.08, 0.085
The default value is currently 0.045. I suggest using instead 0.04.

SeaCelo commented:
@sarvonz Add your review when you have a chance please

rickecon commented:
@SeaCelo and @sarvonz. This PR should have been ready to merge, but it is not. My reform steady state did not solve. This is a very rare occurrence, but I should not be surprised given the test that we ran. The steady state solution can be pretty sensitive to starting values. As such, I need to take the robustness algorithm that I built into OG-Core in [PR #836](https://github.com/PSLmodels/OG-Core/pull/836) and add it to the reform steady state. My failed results from this run are below.

**Baseline steady state equilibrium computation output**
```
GE loop errors =  [-2.508410146262463e-14, -2.5070223674816816e-14, 4.933831121434196e-13, 0.0, 3.461231301571388e-12, -1.1719791803699309e-14, 4.936606678995759e-13, -1.3822276656583199e-14, -7.806255641895632e-15, 2.426600587135397e-13, 5.061367991388011e-13, -5.950101522600448e-16, 1.3844481117075702e-13, -1.092459456231154e-13]
Iteration: 01  Distance:  6.245774321886237e-12
SS debt =  1.061290349047439 0.002900917870000058
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.2494251742773182
Foreign capital holdings =  0.04224643260693746
resource constraint:  [2.01106493e-14]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  7.904787935331115e-14
Maximum error in savings FOC =  4.929390229335695e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_BASELINE/SS/SS_vars.pkl
```

**Baseline transition path equilibrium computation output (20 min, 12 sec)**
```
Maximum debt ratio:  4.903478755422615
w diff:  1.5803858472995813e-06 -5.354729988304996e-07
r diff:  3.21790575019687e-08 -8.971252557499199e-08
r_p diff:  3.21790575019687e-08 -8.971252557499199e-08
p_m diff:  0.0 0.0
BQ diff:  4.3528094994388367e-07 -3.640637166735239e-08
TR diff:  1.0657384290346794e-07 -2.65468051423301e-07
Iteration: 23
	Distance: 8.970078192451272e-06
Max absolute value resource constraint error: 2.1151583345013422e-05
Checking time path for violations of constraints.
Max Euler error, savings:  8.604783552357276e-12
Max Euler error labor supply:  2.9523050670832163e-12
Time path iteration complete.
It took 1211.9648201465607 seconds to get that part done.
run time =  1211.965884923935
```

**Reform steady state equilibrium computation output**
```
GE loop errors =  [0.036488251251174025, 0.049594113119542794, -0.5320786226318877, 0.01524214322893791, -0.17955166455931948, 0.005327985899518087, 0.004273354336397325, -0.0005952612832190347, 0.00723037616235019, 0.024715917993280444, -0.0006519048822944025, 0.012633697617219582, -0.016159649810338753]
Traceback (most recent call last):
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 122, in <module>
    main()
  File "/Users/richardevans/Docs/Economics/OSE/OG-ZAF/./examples/run_og_zaf.py", line 82, in main
    runner(p2, time_path=True, client=client)
  File "/Users/richardevans/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/execute.py", line 45, in runner
    ss_outputs = SS.run_SS(p, client=client)
  File "/Users/richardevans/opt/anaconda3/envs/ogzaf-dev/lib/python3.9/site-packages/ogcore/SS.py", line 1358, in run_SS
    raise RuntimeError("Steady state equilibrium not found")
RuntimeError: Steady state equilibrium not found
```

rickecon commented:
OK. @jdebacker's suggestion fixed it. I updated `run_og_zaf.py`. The reform is now running very well. This PR is ready to merge. Below are my reform steady-state and transition path output that correspond to the baseline run in my previous comment.

**Reform steady state equilibrium computation output**
```
GE loop errors =  [6.938893903907228e-18, 6.938893903907228e-18, 0.0, 0.0, -2.220446049250313e-16, 3.469446951953614e-17, -1.4224732503009818e-15, 4.163336342344337e-17, 3.469446951953614e-17, -1.429412144204889e-15, 2.0816681711721685e-17, 3.660266534311063e-16, -6.938893903907228e-18]
Iteration: 01  Distance:  2.274998632632352e-14
SS debt =  1.269064660476182 0.0034688455944833566
IO:  (1, 1) , C:  (1,)
Foreign debt holdings =  0.298256433211293
Foreign capital holdings =  0.1677431527583257
resource constraint:  [-3.33066907e-15]
Checking constraints on capital, labor, and consumption.
	There were no violations of the constraints on labor  supply.
	There were no violations of the constraints on  consumption.
Maximum error in labor FOC =  7.949196856316121e-14
Maximum error in savings FOC =  4.796163466380676e-14
JUST SAVED SS output to  /Users/richardevans/Docs/Economics/OSE/OG-ZAF/examples/OG-ZAF-Example/OUTPUT_REFORM/SS/SS_vars.pkl
```

**Reform transition path equilibrium computation output (21 min, 25 sec)**
```
Maximum debt ratio:  4.87467226523405
w diff:  1.5875761869565963e-06 -3.954484590895646e-07
r diff:  2.3654958110352897e-08 -9.037962634544927e-08
r_p diff:  2.3654958110352897e-08 -9.037962634544927e-08
p_m diff:  0.0 0.0
BQ diff:  4.332964183076382e-07 -3.657225344094739e-08
TR diff:  1.0994371051759044e-07 -2.6346526710296914e-07
Iteration: 23
	Distance: 8.94115225742335e-06
Max absolute value resource constraint error: 2.1143231016623742e-05
Checking time path for violations of constraints.
Max Euler error, savings:  8.690714814463263e-12
Max Euler error labor supply:  2.97939450888407e-12
Time path iteration complete.
It took 1199.5632781982422 seconds to get that part done.
run time =  1199.5642490386963
Percentage changes in aggregates: Year                    Variable  2021  2022  2023  2024  ...  2028  2029  2030  2021-2030    SS
0                    GDP ($Y_t$) -0.20 -0.20 -0.20 -0.20  ... -0.20 -0.20 -0.19      -0.20 -0.35
1            Consumption ($C_t$) -0.08 -0.13 -0.13 -0.11  ... -0.11 -0.11 -0.11      -0.11 -0.17
2          Capital Stock ($K_t$) -0.42 -0.42 -0.43 -0.43  ... -0.42 -0.42 -0.42      -0.42 -0.76
3                  Labor ($L_t$) -0.02 -0.02 -0.02 -0.02  ... -0.02 -0.02 -0.02      -0.02 -0.03
4     Real interest rate ($r_t$) -0.06 -0.06 -0.06 -0.06  ... -0.07 -0.07 -0.07      -0.06  0.34
5                      Wage rate -0.18 -0.18 -0.18 -0.18  ... -0.18 -0.18 -0.18      -0.18 -0.32

[6 rows x 13 columns]
```

cc: @SeaCelo @sarvonz

rickecon commented:
@SeaCelo. Make sure you updated your `ogzaf-dev` conda environment. Your old environment might have the old version of OG-Core installed from pip.
* `conda remove --name ogzaf-dev --all`
* While in the OG-ZAF directory, re-create the conda environment: `conda env create -f environment.yml`
* Activate the `ogzaf-dev` environment: `conda activate ogzaf-dev`
* Install the `ogzaf` package in this environment: `pip install -e .`

Now run `run_og_zaf.py` from your version of my `demog` branch. Make sure that your version of my `demog` branch was pulled yesterday afternoon after my last commit.

If this doesn't work, let's set up a video call for sometime today for me to look at what you have. You and I have the same hardware setup. This should be working.

SeaCelo commented:
Re-installing OG-CORE solved the issue and I have identical results. Merging this PR
